### ISSUE TITLE
Yolov8 training strategy

### ISSUE THREAD

**Issue by yeonhyochoi** [ultralytics/ultralytics#18809]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I've been developing a detection model for custom dataset using Yolo for months, and I get too much FP.
1. Why Mosaic augmentation is performed first and 'Albumentation' is applied to the dataset where Mosaic is performed
2. If 500,000 images are learned (4 Class, 0: 400,000 main detect objects), appropriate values of Epoch and lr0, lrf parameters
3. If you train 500,000 images, the recommended model size, imgsize, batch size

I trained like this.
epoch : 300
lr0, lrf : 0.01, 0.01
batch : 128
imgsize : 640
model : v8 small or medium
mosaic : 1.0
optimizer : SGD

Here's how I'm doing right now to reduce false positve.
1. Learn the model only with data that can define Class (label exists, human did)
2. Performing Prediction on NormalSet
3. The png where the object is detected in Normalset is defined as false positive, recording all the coordinates of the ROI(model did)
4. Define the data set as a new class called FP and add it to the data set

When proceeding like this, I wonder how to put the background image, which is the method you proposed to reduce FP.

Even after months of experiments, it is difficult to achieve the desired results in Domain (there are many cases where object and background are difficult to distinguish) so I leave an inquiry.

Thank you always.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @yeonhyochoi, thank you for reaching out and sharing your interest in YOLOv8 🚀! Your observations and detailed methodology for reducing false positives (FP) are insightful and much appreciated.

We recommend checking out the official [Docs](https://docs.ultralytics.com/) for a wealth of resources and guidance. For your queries on augmentation strategies, training parameters, and techniques to improve performance, the following documents may be especially relevant:

- [Training Tips](https://docs.ultralytics.com/guides/model-training-tips/) for optimizing your training outcomes
- [Augmentations](https://docs.ultralytics.com/config/#augmentation) to understand how Mosaic and Albumentations are handled
- [Hyperparameters](https://docs.ultralytics.com/config/#hyperparameters) to fine-tune values like `lr0`, `lrf`, and more

Regarding your efforts to reduce false positives, your approach is quite creative, particularly in defining and integrating a new class for FP detection. To assist more effectively, could you provide additional details, such as:

1. Training logs or metrics that illustrate FP rates or trends.
2. Representative examples of problematic images (both FPs and TPs).
3. A [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) of your training setup or script.

This information will help us better understand your situation and offer tailored advice.

In terms of environments, it's essential to ensure you're using the latest version of the `ultralytics` package with up-to-date requirements. Upgrade as follows:

```bash
pip install -U ultralytics
```

Additionally, if you'd like to experiment further, you can access pre-configured environments with GPUs for smoother training:

- **Notebooks** with free GPU:  
  <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a>  
  <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a>  
  <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>

Finally, feel free to engage with the broader Ultralytics community for insights and support. You can join us on [Discord](https://discord.com/invite/ultralytics) 🎧 for real-time discussions, contribute to ongoing threads on [Discourse](https://community.ultralytics.com/), or explore our [Subreddit](https://reddit.com/r/Ultralytics).

📌 **Note:** This is an automated response to provide initial assistance, but rest assured that one of our Ultralytics engineers will review your issue and provide further support as needed.

**Comment by glenn-jocher** [comment]
Thank you for your detailed inquiry. To address your points:

1. Mosaic augmentation is applied first to create composite images that enhance object variety and context, followed by Albumentation for additional transformations. This sequence boosts generalization by diversifying input data.
2. For a large dataset like yours (500,000 images), your current settings (epoch: 300, lr0: 0.01, lrf: 0.01) are reasonable starting points. Consider longer training (e.g., 600 epochs) if underfitting persists, and adjust learning rates based on loss trends and plateau behavior.
3. With 500,000 images and batch size 128, ensure memory sufficiency. For balanced performance, `imgsize: 640` works well, but consider larger sizes (e.g., 1280) if small objects dominate. YOLOv8m or YOLOv8l may perform better for complex datasets.

Regarding background images, add non-target images without labels to your dataset (~5–10% of total) to help reduce false positives. This prevents the model from overfitting to non-object regions. For more tips, refer to [Tips for Best Training Results](https://docs.ultralytics.com/yolov5/tutorials/tips_for_best_training_results/).

Feel free to share training plots or metrics for further insights!

**Comment by yeonhyochoi** [comment]
Thank you for your reply.

Then I would like to ask you an additional question about augmentation.
1. What do you think about learning by applying augmentation to each image and Mosaic in Default setting as follows?
i didnt apply hsv, or flip augmentation in code. because hsv or flip also applied in albumentations. Because if an augmentation like colorjitter is applied first in the albumination and then in the next step, hsv augmentation, the image seems to have a very different image color space than the actual data.

![Image](https://github.com/user-attachments/assets/6878c33c-b0dc-4220-ad65-19b7bca9f97e)

2. The second question is, when you check the saved train_batch, it looks like the image was cut into each cell.
![Image](https://github.com/user-attachments/assets/fa2ccf40-1c00-4f46-b9d1-cb752d418c34)


![Image](https://github.com/user-attachments/assets/f7cbd7f9-eb15-4ab4-879c-2dce081a346b)

I would like to ask if each Mosaic cell I think should be stored as shown in the following figure. Could this problem be caused by the different sizes of input images? Currently, the train data is stored in various sizes of input images and is learning depending on the size set in the train yaml file..

 think that false positiv is caused by the training data that looks like it's been cut off. In fact, false positiv happens a lot on the outside of the image. how do you think.

Thank you always for your kind reply.

**Comment by glenn-jocher** [comment]
@yeonhyochoi thank you for the detailed follow-up. Here are my responses:

1. Using both Albumentations and default YOLO augmentations (e.g., HSV) is generally acceptable, as long as the transformations do not distort the data beyond realistic scenarios. If you feel the resulting image color space is unrealistic, you can adjust the augmentation parameters or disable redundant transformations (like HSV in YOLO) to maintain closer alignment with your dataset's characteristics.

2. The mosaic "cut-off" effect you observe is expected since mosaic augmentation stitches images into quadrants, potentially cropping parts of objects or backgrounds. This is not inherently problematic and can improve generalization. However, if input image sizes vary significantly, resizing during mosaic creation may introduce distortions. To address this, ensure consistent image dimensions either by preprocessing or enabling `rect=True` during training. Additionally, including background images and refining augmentation strategies (e.g., disabling `mosaic` in the final epochs with `close_mosaic` parameter) can help mitigate false positives, particularly at image edges.

For further details on mosaic augmentation, you can refer to [this guide](https://docs.ultralytics.com/usage/cfg/#augmentation-settings). Let me know if you have more questions!

**Comment by yeonhyochoi** [comment]
thanks for your reply.

1 . I understand that the 'rect = True' option you mentioned is a setting that does not use MOSAIC, is it correct?
2. This is an additional question, is there a way to draw a Heatmap during the prediction results??

thanks for your attentions.

**Comment by yeonhyochoi** [comment]
> thanks for your reply.
> 
> 1 . I understand that the 'rect = True' option you mentioned is a setting that does not use MOSAIC, is it correct? 
2. This is an additional question, is there a way to draw a Heatmap during the prediction results?? 
3. Additionally, how do you think about training by changing the 'backgroundset' at specific epochs, such as every epoch or every 5 epochs? Of course, will have to modify the train code . when i modify dataloader, it feel very difficult to me when edit codes. is there any way to change only background imageset? i have a lot of background images, but you recommend 0~10% background images in one epochs. so i want to change background image whenever update epochs.

**Comment by glenn-jocher** [comment]
1. Yes, setting `rect=True` disables Mosaic augmentation by enabling rectangular training, where images are padded to minimize resizing distortion.  
2. To generate a heatmap from prediction results, you can modify the visualization pipeline or use feature maps extracted from the model. Refer to the `masks` function in the [plotting utils](https://docs.ultralytics.com/reference/utils/plotting/) for guidance.  
3. For dynamically updating background images, you would need custom logic in the dataloader to shuffle or replace background samples at specific intervals. This requires modifying the dataset logic, which can be complex, but using techniques like dataset wrappers or hooks may simplify the process. Let us know if you need further clarification.

**Comment by Choiyeonhyo** [comment]
which one is better way you think
1. modify dataloader ans give more background images and select random normalsets whenever train next epoch
2. use weighted random sampler.
in honestly, i want to train models like this
positive 500,000frames / normal 1,000,000frames !! but as you mentioned, too many background images(more than 10%) can make worse results
thanks for your reply

**Comment by glenn-jocher** [comment]
Thank you for your question. Option 1, modifying the dataloader to include random background images per epoch, is preferable as it ensures better dataset diversity and dynamic inclusion of backgrounds without overwhelming the model. Limiting background images to ~10% is essential to avoid overfitting to non-object regions, as too many backgrounds can degrade performance. Let me know if you need further clarification!

**Comment by yeonhyochoi** [comment]
thanks, i`m not sure where to fix.
1. first, Do I really need to use labels.cache in the train data path? In my opinion, it is difficult to use the labels.cache because the dataset changes for each epoch. Or, it is questionable whether the entire data should be stored in the cache and then sliced ​​before use.
2. The code below here seems to be the easiest part to modify in my opinion. in 'trainer' 

![Image](https://github.com/user-attachments/assets/5a46d7db-5683-488c-a038-650ad0365e3f)

there are the class information of whole dataset like this in **'self.train_loader.dataset.labels'**

![Image](https://github.com/user-attachments/assets/64f4ba04-1299-4150-bb49-28318d229288)

then slicing like this
self.train_loader.dataset.labels = self.train_loader.dataset.labels[specificnum:specificnum]
of course, **'self.train_loader.dataset.label_files'** and **'self.train_loader.dataset.im_files'** , too.

thanks for your quick reply !!

**Comment by glenn-jocher** [comment]
You do not need to use `labels.cache` if your dataset changes dynamically each epoch, as caching assumes a static dataset. Instead, your suggested approach of slicing `self.train_loader.dataset.labels`, `label_files`, and `im_files` dynamically before each epoch is valid and aligns with your goal. Modify the dataloader logic carefully to ensure consistency between labels and images. Let me know if you need further clarification.

**Comment by yeonhyochoi** [comment]
thanks a lot. 
This will probably be the last question. As I left the question above, even though I updated all of the data sets, there is a problem where the number of batches actually learned continues to be fixed. Which part needs to be modified further? I could write a custom dataloader myself, but since your version of YOLO is constantly updated, I'll try to use your original code as much as possible.

 i modify like this

![Image](https://github.com/user-attachments/assets/6961bcc0-d2da-4ec7-a434-990997ce13cf)

![Image](https://github.com/user-attachments/assets/1c1a026b-321d-4f2e-a09a-32b516091831)


and , i checked the number of 'nb' 

![Image](https://github.com/user-attachments/assets/98775c1f-130b-4392-b54f-7317cc8e32b2)

However, I checked that the number of batches actually learned is fixed to 22.

![Image](https://github.com/user-attachments/assets/3917210d-99b9-497b-9310-7abda70e1927)

I just wanted to take a moment to sincerely thank you for your patience and for providing such detailed answers to all of my questions. Your insights and support have been invaluable, and I truly appreciate the time and effort you’ve dedicated to helping me. It means a lot to me, and I’m incredibly grateful for your assistance!!!

**Comment by yeonhyochoi** [comment]
additionally, i tried to load InfinitLoader and dynamically edit sampler like this.
but, the speed of train is very slow, and stop train before end of second epoch.


![Image](https://github.com/user-attachments/assets/42b94fc5-932f-437f-9cf5-88102ad9a58d)

**Comment by Y-T-G** [comment]
You can check an example of changing dataloader here.

https://y-t-g.github.io/tutorials/yolo-class-balancing/
